File: leetcode/algorithms/c#/Symmetric Tree/Symmetric Tree.cs
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int x) { val = x; }
 * }
 */
public class Solution
{
    public bool IsSymmetric(TreeNode root)
    {
        if (root == null) return true;
        return dfs(root.left, root.right);
    }

    public bool dfs(TreeNode p, TreeNode q)
    {
        if (p == null && q != null) return false;
        if (p != null && q == null) return false;
        if (p == null && q == null) return true;

        if (p.val != q.val) return false;
        return dfs(p.left, q.right) && dfs(p.right, q.left);
    }
}

File: leetcode/algorithms/c++/Symmetric Tree/Symmetric Tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution
{
private:
    bool sym(TreeNode *x, TreeNode *y)
    {
        if (x == NULL && y != NULL)
            return false;
        if (x != NULL && y == NULL)
            return false;
        if (x == NULL && y == NULL)
            return true;

        if (x->val != y->val)
            return false;
        return sym(x->left, y->right) && sym(x->right, y->left);
    }

public:
    bool isSymmetric(TreeNode *root)
    {
        if (root == NULL)
            return true;
        return sym(root->left, root->right);
    }
};